hypothesis-python/src/hypothesis/strategies/_internal/strategies.py

```diff
@@ -737,15 +736,34 @@ def __repr__(self):
 
     def do_validate(self):
         self.filtered_strategy.validate()
+        fresh = self.filtered_strategy
+        for cond in self.flat_conditions:
+            fresh = fresh.filter(cond)
+        if isinstance(fresh, FilteredStrategy):
+            FilteredStrategy.__init__(
+                self, fresh.filtered_strategy, fresh.flat_conditions
+            )
+        else:
+            FilteredStrategy.__init__(self, fresh, ())
+
+    def filter(self, condition):
+        # Allow strategy rewriting to 'see through' an unhandled predicate.
+        out = self.filtered_strategy.filter(condition)
+        if isinstance(out, FilteredStrategy):
+            return FilteredStrategy(
+                out.filtered_strategy, self.flat_conditions + out.flat_conditions
+            )
+        return FilteredStrategy(out, self.flat_conditions)
```

Comment:
Please add comments here. The flow here is hard to follow, especially since the first thing that `.filter()` does is call `.filter()`..